DSCResources/MSFT_xHostsFile/MSFT_xHostsFile.psm1

```diff
@@ -0,0 +1,267 @@
+            } else {
+                Write-Verbose $localizedData.hostsFileEntryNotFound
+                return $true
+            }
+        }
+    } catch {
+        $exception = $_
+        Write-Verbose ($LocalizedData.anErrorOccurred -f $name, $exception.message)
+        while ($null -ne $exception.innerException) {
+            $exception = $exception.innerException
+            Write-Verbose ($LocalizedData.innerException -f $name, $exception.message)
+        }
+    }
+}
+
+function Test-HostEntry
+{
+    param (
+        [string] $IPAddress,
+        [string] $HostName
+    )
+
+    foreach ($line in Get-Content $script:HostsFilePath) {
+        $parsed = Parse-EntryLine -Line $line
+        if ($parsed.IPAddress -eq $IPAddress) {
+            return $parsed.HostNames -contains $HostName
+        }
+    }
+
+    return $false
+}
+
+function Add-HostEntry
+{
+    param (
+        [string] $IPAddress,
+        [string] $HostName
+    )
+
+    $content = @(Get-Content $script:HostsFilePath)
+    $length = $content.Length
+
+    $foundMatch = $false
+    $dirty = $false
+
+    for ($i = 0; $i -lt $length; $i++) {
+        $parsed = Parse-EntryLine -Line $content[$i]
+
+        if ($parsed.IPAddress -ne $IPAddress) { continue }
+        
+        $foundMatch = $true
+
+        if ($parsed.HostNames -notcontains $HostName) {
+            $parsed.HostNames += $HostName
+            $content[$i] = Reconstruct-Line -ParsedLine $parsed
+            $dirty = $true
+            # Hosts files shouldn't strictly have the same IP address on multiple lines; should we just break here?
+            # Or is it better to search for all matching lines in a malformed file, and modify all of them?
+        }
+    }
+
+    if (-not $foundMatch) {
+        $content += "$IPAddress $HostName"
+        $dirty = $true
+    }
+
+    if ($dirty) {
+        Set-Content $script:HostsFilePath -Value $content
+    }
+}
+
+function Remove-HostEntry
+{
+    param (
+        [string] $IPAddress,
+        [string] $HostName
+    )
+
+    $content = @(Get-Content $script:HostsFilePath)
+    $length = $content.Length
+
+    $placeholder = New-Object psobject
+    $dirty = $false
+
+    for ($i = 0; $i -lt $length; $i++) {
+        $parsed = Parse-EntryLine -Line $content[$i]
+
+        if ($parsed.IPAddress -ne $IPAddress) { continue }
+        
+        if ($parsed.HostNames -contains $HostName) {
+            $dirty = $true
+
+            if ($parsed.HostNames.Count -eq 1) {
+                # We're removing the only HostName from this line; just remove the whole line
+                $content[$i] = $placeholder
+            } else {
+                $parsed.HostNames = $parsed.HostNames -ne $HostName
+                $content[$i] = Reconstruct-Line -ParsedLine $parsed
+            }
+        }
+    }
+
+    if ($dirty) {
+        $content = $content -ne $placeholder
+        Set-Content $script:HostsFilePath -Value $content
+    }
+}
+
+function Parse-EntryLine
```

Comment:
This will need to be changed to an approved verb. I'm open to suggestions on this, maybe read-entryline, or convert-entryline (ie converting it into an  custom object).